src/search/pdbs/cegar.cc

```diff
@@ -0,0 +1,589 @@
+    }
+    return flaws;
+}
+
+FlawList CEGAR::get_flaws() {
+    FlawList flaws;
+    State concrete_init = task_proxy.get_initial_state();
+    for (size_t collection_index = 0;
+         collection_index < projection_collection.size(); ++collection_index) {
+        if (projection_collection[collection_index] &&
+            !projection_collection[collection_index]->is_solved()) {
+            FlawList new_flaws =
+                get_flaws_for_projection(collection_index, concrete_init);
+            if (concrete_solution_index != -1) {
+                assert(concrete_solution_index == static_cast<int>(collection_index));
+                assert(new_flaws.empty());
+                // Return empty flaws to terminate the algorithm.
+                flaws.clear();
+                return flaws;
+            }
+            for (Flaw &flaw : new_flaws) {
+                flaws.push_back(move(flaw));
+            }
+        }
+    }
+    return flaws;
+}
+
+void CEGAR::add_pattern_for_var(int var) {
+    projection_collection.push_back(
+        compute_projection(
+            task, {var}, rng, wildcard_plans, verbosity));
+    variable_to_projection[var] = projection_collection.size() - 1;
+    collection_size += projection_collection.back()->get_pdb()->get_size();
+}
+
+bool CEGAR::can_merge_patterns(int index1, int index2) const {
+    int pdb_size1 = projection_collection[index1]->get_pdb()->get_size();
+    int pdb_size2 = projection_collection[index2]->get_pdb()->get_size();
+    if (!utils::is_product_within_limit(pdb_size1, pdb_size2, max_pdb_size)) {
+        return false;
+    }
+    int added_size = pdb_size1 * pdb_size2 - pdb_size1 - pdb_size2;
+    return collection_size + added_size <= max_collection_size;
+}
+
+void CEGAR::merge_patterns(int index1, int index2) {
+    // Merge projection at index2 into projection at index2.
+    Projection &projection1 = *projection_collection[index1];
+    Projection &projection2 = *projection_collection[index2];
+
+    const Pattern &pattern2 = projection2.get_pattern();
+    for (int var : pattern2) {
+        variable_to_projection[var] = index1;
+    }
+
+    // Compute merged pattern.
+    Pattern new_pattern = projection1.get_pattern();
+    new_pattern.insert(new_pattern.end(), pattern2.begin(), pattern2.end());
+    sort(new_pattern.begin(), new_pattern.end());
+
+    // Store old pdb sizes.
+    int pdb_size1 = projection_collection[index1]->get_pdb()->get_size();
+    int pdb_size2 = projection_collection[index2]->get_pdb()->get_size();
+
+    // Compute merged projection.
+    unique_ptr<Projection> merged =
+        compute_projection(
+            task, move(new_pattern), rng, wildcard_plans, verbosity);
+
+    // Update collection size.
+    collection_size -= pdb_size1;
+    collection_size -= pdb_size2;
+    collection_size += merged->get_pdb()->get_size();
+
+    // Clean up.
+    projection_collection[index1] = move(merged);
+    projection_collection[index2] = nullptr;
```

Comment:
That's right, but I don't think gabs are a problem here. It makes iterating over the collection a bit more tedious, but we could actually write an iterator for that. I don't like using shared_ptr for something which is not shared only to avoid gaps. I'd rather try to hide the gaps behind an iterator.